libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1323,30 +1302,62 @@ void IRGeneratorForStatements::appendExternalFunctionCall(
 			if iszero(extcodesize(<address>)) { revert(0, 0) }
 		</checkExistence>
 
+		// storage for returned data and arguments
 		let <pos> := <freeMemory>
-
 		mstore(<pos>, <shl28>(<funId>))
 		let <end> := <encodeArgs>(add(<pos>, 4) <argumentString>)
 
 		let <result> := <call>(<gas>, <address>, <?hasValue> <value>, </hasValue> <pos>, sub(<end>, <pos>), <pos>, <reservedReturnSize>)
-		if iszero(<result>) { <forwardingRevert>() }
-
-		<?dynamicReturnSize>
-			returndatacopy(<pos>, 0, returndatasize())
-		</dynamicReturnSize>
-
-		mstore(<freeMemoryPointer>, add(<pos>, and(add(<returnSize>, 0x1f), not(0x1f))))
-		<?returns> let <retVars> := </returns> <abiDecode>(<pos>, add(<pos>, <returnSize>))
+		<?isTryCall>
+			let <trySuccessConditionVariable> := <result>
+		<!isTryCall>
+			if iszero(<result>) { <forwardingRevert>() }
+			<?hasRetVars>
+				let <retVars> := <decodeReturnParameters>(<pos>, <returnSize>)
```

Comment:
I think we have a problem here:
If we have static return data, we still use `returndatasize()` but we do not use `returndatacopy`. So if the call returns more data than expected, we could read invalid data.

Reply:
Sorry, I think this is not a problem after all:
 - if the contract returned more data than expected, we will only read the expected data in the static case
 - if the contract returned less data, the decoder will throw
 - in the dynamic case, we use `returndatacopy`, so the two sizes match.